feat(dev): add Conventional Commit Pipeline Script

WarpWing opened the pull request:
**Pull Request Description**

Implements a squash merge script with conventional commit formatting as specified in #126

Closes #126 

**Licensing Certification**

FarmData2 is a [Free Cultural Work](https://freedomdefined.org/Definition) and all accepted contributions are licensed as described in the LICENSE.md file. This requires that the contributor holds the rights to do so. By submitting this pull request **I certify that I satisfy the terms of the [Developer Certificate of Origin](https://developercertificate.org/)** for its contents.

WarpWing committed:
Create convencommitcheck.bash

Signed-off-by: Ty Chermsirivatana <professornitro@gmail.com>

WarpWing commented:
It's not perfect but it's what I have in mind. It's a shell but I hope it provides good standing ground for others to perhaps work upon?

WarpWing committed:
Merge branch 'development' into patch-4

WarpWing committed:
Added squashMerge functionality

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Added GH CLI Login Functionality

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Update convencommitcheck.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Update convencommitcheck.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing commented:
@braughtg I've created a mockup of what was discussed in #126 (as I understand it hopefully) with some primitive functionality with squash merging, GitHub CLI Authentication checks, and some level of creating a conventional commit pipeline script. Would appreciate your thoughts.

WarpWing committed:
Update convencommitcheck.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

braughtg commented on bin/convencommitcheck.bash:
```diff
@@ -0,0 +1,143 @@
+    local type=$(echo "$pr_title" | cut -d':' -f1)
+    local description="$pr_title"
+
+    # Default scope to 'core' if not provided
```
I think having the default be no scope at all would be fine.  So no `(...)` in the message, just the type.

WarpWing replied:
I see, I was curious what no scope would be. I'll make the changes in my next iteration.

braughtg reviewed (changes requested):
This looks like it's heading in a good direction.  

A few things that come to mind for additional improvement:
- It will need to work if the PR title is currently a conventional commit or if it is not.  For example, sometimes I create my PR's with a title that is a conventional commit message so that I don't have to edit it later.  But generally when others create a PR it is not expected that the title will be a conventional commit.
- It would be good to also include the option of making the commit be a `BREAKING` change (see: https://www.conventionalcommits.org/en/v1.0.0/).  But also note that the sematic release bot I use doesn't honor the `!` in the message, so we need to add the `BREAKING CHANGE:` footer.  If this footer exists in the PR body then it should be pulled into the squash commit message footer.
- it would be great to clean this up so that it guides the maintainer through a series of prompts that produce the end result.  So things like:
  - Prompt for PR number if not provided 
  - Prompt for type / scope if they do not appear in the PR title.
  - Prompt with proposed conventional commit with options for accept (A) or edit (E).
  - Prompt for whether change is a "BREAKING CHANGE:" if it doesn't appear int the body of the PR, and for accept (A) or edit (E) if it does appear.
  - Prompt with proposed body of commit as extracted from body of PR, including any "BREAKING CHANGE:" footer with options for accept (A) or edit (E).
  - Provide a final opportunity to accept or edit.

Maybe choosing to edit opens the title/message in `nano`?

WarpWing committed:
Added Suggested Changes

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing commented:
Finally implemented the draft changes I made this afternoon in response to your feedback and fixed the quick defaulting core scope issue.

WarpWing committed:
Merge branch 'development' into patch-4

WarpWing commented:
@braughtg I never got to work on this further (as well as Selective Doc builds) but outside of this, I was thinking if we want the script to use default env editor or give a list instead of just nano. I use nano personally so it doesn't matter at least to me, just curious. Also, is there anything else you would like me to add maybe any cli or options features?

braughtg commented:
> if we want the script to use default env editor or give a list instead of just nano.

I'm happy just having it use nano.

WarpWing committed:
Merge branch 'development' into patch-4

WarpWing committed:
Restructured formatting

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Update convencommitcheck.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

braughtg commented on bin/convencommitcheck.bash:
```diff
@@ -0,0 +1,109 @@
+# Validate and prompt for type and scope
+TYPE=${TYPE:-$(promptForValue "Enter commit type" "feat")}
+if ! elementInArray "$TYPE" "${VALID_TYPES[@]}"; then
+    echo "Invalid commit type: $TYPE. Please enter a valid type."
```
Might be nice to list the valid types here to save the step of going and looking them up.  I know I always seem to forget exactly what they are when I'm not using them every day.

braughtg commented on bin/convencommitcheck.bash:
```diff
@@ -0,0 +1,109 @@
+fi
+SCOPE=${SCOPE:-$(promptForValue "Enter commit scope" "none")}
+if ! elementInArray "$SCOPE" "${VALID_SCOPES[@]}"; then
+    echo "Invalid commit scope: $SCOPE. Please enter a valid scope."
```
Same here... Listing the valid scopes would be helpful.  Maybe just print a header and then create a `printArray` method that displays them neatly.

WarpWing replied:
Will get on that. Before I do, if it isn't a valid scope, should I just start from the beginning of the script again(and make them input a valid scope) or just exit and make the user start over again?

braughtg replied:
In general, I'd say check each input as soon as they enter it and if it is invalid, then prompt again.  It shouldn't be necessary to start over from the beginning of the script.

braughtg commented on bin/convencommitcheck.bash:
```diff
@@ -0,0 +1,109 @@
+elementInArray() {
+    local element
+    for element in "${@:2}"; do
+        [[ "$element" == "$1" ]] && return 0
```
Technically for conventional commits these comparisons should be case insensitive.

Here's more information: https://www.conventionalcommits.org/en/v1.0.0/#specification

WarpWing replied:
Intergrated shopt for case insensitive matching during script runtime (turns it off after)

braughtg commented on bin/convencommitcheck.bash:
```diff
@@ -0,0 +1,109 @@
+    fi
+    commit_message="${commit_message}: ${description}"
+    if [[ "$breaking_change" == "yes" ]]; then
+        commit_message="${commit_message}\n\nBREAKING CHANGE: ${body}"
```
Open to discussion:  I don't know if we want the entire body to become the BREAKING CHANGE description.  It seems likely that some PRs will have both non-breaking changes and BREAKING CHANGES to describe.  So maybe we should have a way to pull those apart?  Not sure what it would be.  Any thoughts?

WarpWing replied:
> Open to discussion: I don't know if we want the entire body to become the BREAKING CHANGE description. Some PRs will likely have both non-breaking changes and BREAKING CHANGES to describe. So we should have a way to pull those apart. Not sure what it would be. Any thoughts?

That's an interesting one. My thought is to implement an interactive approach that prompts the author to manually input the breaking change description if a breaking change is indicated and putting it in the commit body is my recommendation. Something along the lines of this (We would also have to modify `convertToConventionalCommit()` to accept new parameters as well):
```bash
BREAKING_CHANGE=$(promptForValue "Is this a breaking change (yes/no)" "no")
BREAKING_CHANGE_DESCRIPTION=""
if [[ "$BREAKING_CHANGE" == "yes" ]]; then
    echo "You indicated this is a breaking change. Please provide a specific description of the breaking change."
    read -p "Enter breaking change description: " BREAKING_CHANGE_DESCRIPTION
fi
```
The way that I see it, if it has any amount of breaking changes, it should be treated if it was holistically breaking changes only. But, those are my thoughts. Thoughts on the thoughts?

braughtg replied:
> My thought is to implement an interactive approach that prompts the author to manually input the breaking change description if a breaking change is indicated and putting it in the commit body

This sounds like a good approach.  But only if the PR body doesn't already contain a BREAKING CHANGE footer.  If it does, extract it and use that.

braughtg replied:
> The way that I see it, if it has any amount of breaking changes, it should be treated if it was holistically breaking changes only.

Yes.  If a conventional commit has a body and a BREAKING CHANGE footer, it will bump the major version number.  So if a commit contains both non-breaking changes and breaking changes, the semantic version number will be adjusted as a BREAKING CHANGE.

WarpWing replied:
> > The way that I see it, if it has any amount of breaking changes, it should be treated if it was holistically breaking changes only.
> 
> Yes. If a conventional commit has a body and a BREAKING CHANGE footer, it will bump the major version number. So if a commit contains both non-breaking changes and breaking changes, the semantic version number will be adjusted as a BREAKING CHANGE.

Perfect! I assume we would then adhere to something like this
```
<type>(<scope>): <short description>

<BODY>

BREAKING CHANGE: <breaking change description>
```

WarpWing replied:
Added BREAKING CHANGE checks in PR Body

braughtg commented on bin/convencommitcheck.bash:
```diff
@@ -0,0 +1,109 @@
+TYPE=$(echo "$PR_TITLE" | cut -d':' -f1)
+SCOPE=$(echo "$PR_TITLE" | cut -d'(' -f2 | cut -d')' -f1)
+DESCRIPTION=$(echo "$PR_TITLE" | cut -d')' -f2- | cut -d':' -f2-)
+
```
Here it is possible that the PR will also include a "BREAKING CHANGE" footer in the description.  If it does, we should pull that out separately and use it.  If that is the case, then we won't prompt below.

WarpWing replied:
I'll wait to finish the open discussion to finalize these changes.

WarpWing replied:
I implemented a check here for breaking changes when prompted.
```bash
if [[ "$BREAKING_CHANGE" == "yes" ]]; then
    # Check if the PR body already contains a BREAKING CHANGE footer and extract it if present in PR body
    if [[ "$PR_BODY" =~ "BREAKING CHANGE:" ]]; then
        BREAKING_CHANGE_DESCRIPTION=$(echo "$PR_BODY" | sed -n '/BREAKING CHANGE:/,$p')
        echo "Breaking changes found and appended to commit."
    else
        echo "You indicated this is a breaking change. Please provide a specific description of the breaking change."
        read -p "Enter breaking change description: " BREAKING_CHANGE_DESCRIPTION
    fi
fi
```

braughtg reviewed (changes requested):
This is starting to look really good!  Thanks for your efforts on it.

I've got a few more suggestions for improvements in the embedded comments and a feature that is open to further discussion.

WarpWing commented:
@braughtg should I go and implement a help command as well? I was also considering going for that and also a cli option that would allow cli flags to input a PR with scope already like 

```bash
convencommitcheck,.bash --pr #89 -- type ci --scope farmdata2 --desc text
```

which then only checks for Github cli authentication and then goes straight to 
```bash
CONV_COMMIT=$(convertToConventionalCommit "$TYPE" "$SCOPE" "$DESCRIPTION" "$PR_BODY" "$BREAKING_CHANGE")
```

After I implement the breaking change functions, I'll look at this next given your thoughts.

braughtg commented:
> @braughtg should I go and implement a help command as well? I was also considering going for that and also a cli option that would allow cli flags to input a PR with scope 

These both sound like nice ideas.  I'd definitely like a `--help` flag that describes the tool, though it won't have much to say if there aren't the other flags for type/scope/title/ etc...  So doing them all together makes sense to me.

WarpWing committed:
Intergrate case insensitive comparision with shopt and updated type and scope to normalize type comparision after extraction and validation

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Rename convencommitcheck.bash to convocommit.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Merge branch 'development' into patch-4

WarpWing committed:
Adding in persistent validation for type and scope input

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Implemented BREAKING CHANGES checks 

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Update convocommit.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing commented:
I'm going to clean up the script tonight, add cli options, and a help command. I'm also going to make a repository flag so this script can be ported to work with any repository (I imagine this tooling would be useful across multiple repositories).

braughtg commented:
> I'm also going to make a repository flag so this script can be ported to work with any repository (I imagine this tooling would be useful across multiple repositories).

I can image this will be useful in testing the script as well.  I'd prefer not to run tests against the main repo and end up having a bunch of extra commits in there just from the testing.

braughtg commented:
A naming suggestion:  Would it make sense to call this script "squashMergePR.bash" or something similar?

WarpWing commented:
> A naming suggestion: Would it make sense to call this script "squashMergePR.bash" or something similar?

 I was tinkering around with the name and your name makes sense. Just had a placeholder temporarily.

WarpWing committed:
Rename convocommit.bash to squashMergePR.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing commented:
> > I'm also going to make a repository flag so this script can be ported to work with any repository (I imagine this tooling would be useful across multiple repositories).
> 
> I can image this will be useful in testing the script as well. I'd prefer not to run tests against the main repo and end up having a bunch of extra commits in there just from the testing.

Should I hardcode FarmData2's repo as the default and give an option to set a flag, prompt the user during the program to set it within the script or do both(obviously bypassing if a cli flag is present)?

braughtg commented:
Can you get the repo url information from the repo with git? Then if the script is moved no changes are necessary.  Having a cli override would be good too.

On Feb 28, 2024, at 19:02, Ty Chermsirivatana ***@***.***> wrote:

﻿
Caution:This email originated from outside of the College. Do not click links or open attachments unless you recognize the sender and know the content is safe. Report anything suspicious to the HelpDesk.


I'm also going to make a repository flag so this script can be ported to work with any repository (I imagine this tooling would be useful across multiple repositories).

I can image this will be useful in testing the script as well. I'd prefer not to run tests against the main repo and end up having a bunch of extra commits in there just from the testing.

Should I hardcode FarmData2's repo as the default and give an option to set a flag, prompt the user during the program to set it within the script or do both(obviously bypassing if a cli flag is present)?

—
Reply to this email directly, view it on GitHub<https://github.com/FarmData2/FarmData2/pull/136#issuecomment-1970134663>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEBNHMRYK4K67Q664VUR7VLYV7AP7AVCNFSM6AAAAABBVMIEGCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZQGEZTINRWGM>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

WarpWing commented:
> Can you get the repo url information from the repo with git? Then if the script is moved no changes are necessary. Having a cli override would be good too. On Feb 28, 2024, at 19:02, Ty Chermsirivatana ***@***.***> wrote: ﻿ Caution:This email originated from outside of the College. Do not click links or open attachments unless you recognize the sender and know the content is safe. Report anything suspicious to the HelpDesk. I'm also going to make a repository flag so this script can be ported to work with any repository (I imagine this tooling would be useful across multiple repositories). I can image this will be useful in testing the script as well. I'd prefer not to run tests against the main repo and end up having a bunch of extra commits in there just from the testing. Should I hardcode FarmData2's repo as the default and give an option to set a flag, prompt the user during the program to set it within the script or do both(obviously bypassing if a cli flag is present)? — Reply to this email directly, view it on GitHub<[#136 (comment)](https://github.com/FarmData2/FarmData2/pull/136#issuecomment-1970134663)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEBNHMRYK4K67Q664VUR7VLYV7AP7AVCNFSM6AAAAABBVMIEGCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZQGEZTINRWGM>. You are receiving this because you were mentioned.Message ID: ***@***.***>

I'm not sure... are you asking if, given an URL I can parse information or if you run the script within the context of the repository could I get information, because I'm sure .git has the information I need (from what I assume I'll need to look into it).

braughtg commented:
> I'm not sure... are you asking if, given an URL I can parse information or if you run the script within the context of the repository could I get information, because I'm sure .git has the information I need (from what I assume I'll need to look into it)

I was suggesting we could get the Repo URL from git. Then use that instead of hard coding it or getting it from the user.  I think just getting the upstream remote or the origin if the upstream is not defined should work.  Maybe have the user confirm? And provide a flag to override.

WarpWing commented:
> I'm not sure... are you asking if, given an URL I can parse information or if you run the script within the context of the repository could I get information, because I'm sure .git has the information I need (from what I assume I'll need to look into it)
> I was suggesting we could get the Repo URL from git. Then use that instead of hard coding it or getting it from the user. I think just getting the upstream remote or the origin if the upstream is not defined should work. Maybe have the user confirm? And provide a flag to override.

I see, one thing I'm not sure how to implement is the PR title and body, which we can surely have the user put in but maybe there's a better way, at least for viewing we can always attach the `https://github.com/FarmData2/FarmData2/pull/` + num but in terms of putting the information, maybe using cURL?

braughtg commented:
> one thing I'm not sure how to implement is the PR title and body, which we can surely have the user put in but maybe there's a better way

I think you already have code in there to grab the PR title and body using `gh` (lines 106-107).

My thinking is that you could use `git` to find the URL of the repo and then use `gh` to get the PR title and body.  Then parse that with `cut` and `grep` to get the information for the user to confirm/edit into the conventional commit message.  If I understand the code you had, most of that is already there.

Or I may be confused about what you are asking here.

WarpWing commented:
> > one thing I'm not sure how to implement is the PR title and body, which we can surely have the user put in but maybe there's a better way
> 
> I think you already have code in there to grab the PR title and body using `gh` (lines 106-107).
> 
> My thinking is that you could use `git` to find the URL of the repo and then use `gh` to get the PR title and body. Then parse that with `cut` and `grep` to get the information for the user to confirm/edit into the conventional commit message. If I understand the code you had, most of that is already there.
> 
> Or I may be confused about what you are asking here.

Ohhh, I totally misunderstood you. I thought you wanted me to replace gh cli with the repo link for some reason. Could you clarify what you mean by repo url information?

WarpWing commented:
I see what you mean, you're talking about getting and defining the repository from a URL but getting the information about PRs from gh. I see, never mind, I can do that.

WarpWing committed:
Update squashMergePR.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Added help command and parsing cli arguments

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Added cli flags, help, git source check, corrected editing, url formatting etc

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Merge branch 'FarmData2:development' into patch-4

WarpWing commented:
After some extensive reformatting and commits, here's what I've added:

- Added command line flags (including examples and supported ingestion of multiple repositories link types)
- Added dependency check to ensure `gh` is installed
- Added in relative git directory to gain auto context for envs
- Created editing for commits by creating a temporary file for modifying commit message
- Refactored and organized functions as well as added a clear entrypoint for readability 
- Implemented help command with `--help`

WarpWing committed:
Update squashMergePR.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Update squashMergePR.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Merge branch 'development' into patch-4

WarpWing committed:
Update squashMergePR.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

braughtg commented:
@WarpWing Overall I really like the direction and the changes you have made. 

I just created a new repo for testing this script.  I sent you an invite to that repo with write permission.

We can use that repo for continued development and testing of the script.  it will allow us to make PRs and try to merge them without adding them to the FarmData2 repo.  Once we have it working, we can put it back into this PR to add it to FarmData2.

The repo is here: https://github.com/FarmData2/squashMergeTest

I opened an issue there reporting on my experience running the script.  Note that I was running it within the FarmData2 development environment (i.e. the Debian linux docker desktop container).

WarpWing commented:
> @WarpWing Overall I really like the direction and the changes you have made.
> 
> I just created a new repo for testing this script. I sent you an invite to that repo with write permission.
> 
> We can use that repo for continued development and testing of the script. it will allow us to make PRs and try to merge them without adding them to the FarmData2 repo. Once we have it working, we can put it back into this PR to add it to FarmData2.
> 
> The repo is here: https://github.com/FarmData2/squashMergeTest
> 
> I opened an issue there reporting on my experience running the script. Note that I was running it within the FarmData2 development environment (i.e. the Debian linux docker desktop container).

Thanks for the issue, I think I see where I went wrong with the gh cli, I think I messed up some of the cli flags and didn't correctly look at the scope and test, let me clone the squashRepo, test it and get an MVP to you soon.
